library/src/scala/tasty/reflect/IsInstanceOf.scala

```diff
@@ -0,0 +1,39 @@
+package scala.tasty
+package reflect
+/* FIXME Using class tags to type tests abstract type members in unsound
+
+   ```
+   trait R {
+     type Nat
+     type Succ <: Nat
+     type Idx
+     given ClassTag[Nat]
+     given ClassTag[Succ]
+     given ClassTag[Idx]
+     def n: Nat
+   }
+   object RI extens R {
+     type Nat = Int
+     type Succ = Int
+     type Idx = Int
+     given ClassTag[Nat] = classOf[Integer]
+     given ClassTag[Succ] = new ClassTag[Integer] {
+       def runtimeClass = classOf[Integer]
+       def unapply(x: Any): Option[Succ] = x match
+         case n: Int if n > 0 => Some(n)
+         case _ => None
+     }
+     given ClassTag[Idx] = classOf[Integer]
+     def n: Nat = 4
+   }
+   val r1: R = RI
+   val r2: R = RI
+   r1.n match {
+     case n: r2.Nat => // Should not match or should have an unchecked waring
+     case n: r1.Idx => // Should not match or should have an unchecked waring
+     case n: r1.Succ => // Should match only if r1.n is an r1.Succ under the constraints set in r1
+   }
+   ```
+*/
```

Comment:
Given comment https://github.com/lampepfl/dotty/pull/7394#issuecomment-553272477, there is the question if it it possible to fix class tag to be sound even when considering paths.

Reply:
@odersky @sjrd WDYT? 

Note that there is also a case we did not discuss with `r1.Nat` and `r1.Idx` where the paths are equal, the classtags are equal and the unapply of the class tag are the same but there is no subtyping relationship between the two abstract types.

Reply:
To me, from the last discussion we had, there is only one reasonable path forward that will solve all the problems at once, with a clean design, and no migration issue:

First, introduce a new typeclass
```scala
trait TypeTest[A] {
  def isInstance(x: Any): Boolean
}
```
that can be synthesized by the compiler for `A = p.T` in exactly the same situations where an `x.isInstanceOf[p.T]` would be valid, and synthesizes it precisely as
```scala
new TypeTest[p.T] {
  def isInstance(x: Any): Boolean = x.isInstanceOf[p.T]
}
```

Then, in pattern-matching, for `case x: A` or `case Extractor(x)` where the `Extractor` expects an `A`, if `A` cannot be tested natively, try summoning a `TypeTest[A]`. Only if that fails as well, try summoning a `ClassTag[A]`.

Finally, deprecated `ClassTag.unapply` and summoning a `ClassTag[A]` to implement a type test, since they are unsound.

`ClassTag`s were designed for `Array`s. This is the use case they are widely used for, and they work *perfectly* for that use case. It is only later on that the type-test feature was bolted on, without giving it more thought than a PR review. Please let's not pretend that `ClassTag`s are broken and need fixing because *that bolted-on feature* is broken. `ClassTag`s are correct. It's using `ClassTag`s for type tests that's broken. The correct fix is not to fix `ClassTag`s but to fix type-tests.

---

The above `TypeTest` typeclass is slightly unsound because one can implement a custom `TypeTest` for any `A` that always returns `true`. The subsequent compiler-generated `.asInstanceOf[A]` will throw a CCE. We can fix that design by complicating a bit the `TypeTest` API as follows:
```scala
trait TypeTest[A] {
  def isInstance(x: Any): TypeTest.Result[x.type & A]
}
object TypeTest {
  opaque type Result[A] = Boolean

  def success[A](x: A): Result[A] = true
  def failure[A]: Result[A] = false
}
```
Now a synthesized `TypeTest[p.T]` would look like
```scala
new TypeTest[p.T] {
  def isInstance(x: Any): TypeTest.Result[x.type & p.T] = x match {
    case x: p.T => TypeTest.success(x)
    case _      => TypeTest.failure
  }
}
```
and we can safely write custom `TypeTest[A]` instances that are sound down the line.

Reply:
Small fix:

```scala
new TypeTest[p.T] {
  def isInstance(x: Any): TypeTest.Result[x.type & p.T] = x match {
    case x: (p.T & x.type) => TypeTest.success(x)
    case _                 => TypeTest.failure
  }
}
```

Reply:
There is still something missing, when we do have an implicit `p2.Y` in scope we can match a `p1.Y` without an unchecked warning.
```scala
object Test {
  def main(args: Array[String]): Unit = {
    val p1: T = T1
    import p1.given

    val p2: T = T1
    import p2.given

    (p1.y: p1.X) match {
      case x: p2.Y => // should be an unchecked cast but is not
      case x: p1.Y =>
      case _ =>
    }
  }

}

trait T {
  type X
  type Y <: X
  def x: X
  def y: Y
  given TypeTest[Y] = typeTestOfY
  protected def typeTestOfY: TypeTest[Y]
}

object T1 extends T {
  type X = Boolean
  type Y = true
  def x: X = false
  def y: Y = true
  protected def typeTestOfY: TypeTest[Y] = new {
    def isInstance(x: Any): TypeTest.Result[x.type & Y] = x match
      case x: (true & x.type) => TypeTest.success(x)
      case _ => TypeTest.failure
  }
}
```


This can be solved by adding adding a second type parameter like in #7394. 

```scala
trait TypeTest[S, T <: S] {

  def isInstance(x: S): TypeTest.Result[x.type & T]

  def unapply(x: S): Option[x.type & T] =
    TypeTest.unapply(x)(given this)

}

object TypeTest {

  opaque type Result[A] = Boolean

  def success[A](x: A): Result[A] = true

  def failure[A]: Result[A] = false

  def unapply[S, T <: S](x: S)(given tt: TypeTest[S, T]): Option[x.type & T] =
    if tt.isInstance(x) then Some(x.asInstanceOf[x.type & T])
    else None

}
```
And using it in the following way
```scala
trait T {
  type X
  type Y <: X
  def x: X
  def y: Y
  given TypeTest[X, Y] = typeTestOfY
  protected def typeTestOfY: TypeTest[X, Y]
}
object T1 extends T {
  type X = Boolean
  type Y = true
  def x: X = false
  def y: Y = true
  protected def typeTestOfY: TypeTest[X, Y] = new {
    def isInstance(x: X): TypeTest.Result[x.type & Y] = x match
      case x: (true & x.type) => TypeTest.success(x)
      case _ => TypeTest.failure
  }
}
```

Though we would need to ensure that `S` in `TypeTest[S, T]` is defined locally to ensure we do not get the previous unsoundness.

Reply:
We could also consider defining it like
```scala
trait TypeTest[S, T <: S] {
  def unapply(x: S): TypeTest.Result[x.type & T]
}

object TypeTest {

  opaque type Result[A] = A | Failure.type

  object Result {
    given [A](res: Result[A]) { // currently fails, not sure where to put the extension methods
      def isEmpty: Boolean = res == Failure
      def get: A = res.asInstanceOf[A]
    }
  }

  private object Failure

  def success[A](x: A): Result[A] = x

  def failure[A]: Result[A] = Failure

}
```

Reply:
Moved discussion to #7554